ocaml/xapi/xapi_xenops.ml

```diff
@@ -2281,13 +2284,25 @@ let update_vm ~__context id =
               try
                 Option.iter
                   (fun (_, state) ->
-                    debug "xenopsd event: Updating VM %s last_start_time <- %s"
-                      id
-                      (Date.to_string (Date.of_float state.Vm.last_start_time)) ;
                     let metrics = Db.VM.get_metrics ~__context ~self in
                     let start_time = Date.of_float state.Vm.last_start_time in
-                    Db.VM_metrics.set_start_time ~__context ~self:metrics
-                      ~value:start_time ;
+                    if
+                      start_time
+                      <> Db.VM_metrics.get_start_time ~__context ~self:metrics
+                    then (
+                      debug
+                        "xenopsd event: Updating VM %s last_start_time <- %s" id
+                        (Date.to_string (Date.of_float state.Vm.last_start_time)) ;
+                      Db.VM_metrics.set_start_time ~__context ~self:metrics
+                        ~value:start_time ;
+                      if
+                        power_state <> `Suspended (* not VM suspend *)
```

Comment:
Would suspending a VM reach here? And what's more, resuming a VM may not reach here also.
This is based on what I read from:
https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/xenops_server_xen.ml#L1433-L1473

Reply:
Yes, based on tests, and this change passed my local test.
From test, when a VM is suspended, the "Suspended" power state changed comes together with "last_start_time" update:

Below is one piece of log:

Dec 22 06:02:16 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: processing event for VM b3587353-cbcf-6008-dc9b-e00038467dbf
Dec 22 06:02:16 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf power_state <- Suspended
Dec 22 06:02:16 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf domid
Dec 22 06:02:16 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf memory_target <- 0
Dec 22 06:02:16 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf hvm <- false
Dec 22 06:02:16 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf last_start_time <- 19700101T00:00:00Z
Dec 22 06:02:16 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf vcpu_target <- 0
Dec 22 06:02:16 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: ignoring event for VBD (VM b3587353-cbcf-6008-dc9b-e00038467dbf not resident)
Dec 22 06:02:16 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: ignoring event for VIF (VM b3587353-cbcf-6008-dc9b-e00038467dbf not resident)

Reply:
OK. For suspend, `last_start_time <- 19700101T00:00:00Z`.
What about resume? The last_start_time will be changed as well?

Reply:
Set to the time when it is resumed:

Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: processing event for VM b3587353-cbcf-6008-dc9b-e00038467dbf
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf current_domain_type <- hvm
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf power_state <- Running
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf domid
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf consoles
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf memory_target <- 4294967296
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf platform:timeoffset <- 0
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf hvm <- true
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf nomigrate <- false
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf nested_virt <- false
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf last_start_time <- 20231222T05:59:07Z
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: VM b3587353-cbcf-6008-dc9b-e00038467dbf domid 59 uncooperative_balloon_driver = false
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf domid 59 guest_agent
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf PV drivers detected false
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf domid 59 xsdata
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf domid 59 memory target
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf vcpu_target <- 2
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf shadow_multiplier <- 1.00
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VM b3587353-cbcf-6008-dc9b-e00038467dbf last_boot_CPU_flags <- 1fcbfbff-f7ba2223-2c100800-00000001-00000001-00000281-00000000-00000000-00001000-9c000400-00000000-00000000-00000000-00000000-00000000-00000000-00000000-00000000-00000000-00000000-00000000-00000000
Dec 22 06:14:12 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VBD b3587353-cbcf-6008-dc9b-e00038467dbf.xvda device <- xvda; currently_attached <- true
Dec 22 06:14:13 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenopsd event: Updating VIF b3587353-cbcf-6008-dc9b-e00038467dbf.0 currently_attached <- true
Dec 22 06:14:13 xrtmia-12-07 xapi: [debug||88 |org.xen.xapi.xenops.classic events D:642e31003c24|xenops] xenops event on VM b3587353-cbcf-6008-dc9b-e00038467dbf

Reply:
>Set to the time when it is resumed:

I don't think so. The `last_start_time` value is persisted in the xenopsd metadata record and saved in the `VM.last_booted_record` field in xapi. On resume, this xenopsd metadata is restored in xenopsd. So `last_start_time` should still be the actual last start time.

Your example has:

    Dec 22 06:14:12 (...) last_start_time <- 20231222T05:59:07Z

which is a `last_start_time` 15 min before this is being logged. I assume this is when you had started the VM.

Reply:
It doesn't matter for this patch though.

Reply:
Oh, I didn't notice that, thank you!